src/mission_companion.cpp

```diff
@@ -1420,23 +3400,128 @@ void talk_function::companion_return( npc &comp ){
     assert( !comp.is_active() );
     comp.reset_companion_mission();
     comp.companion_mission_time = calendar::before_time_starts;
+    comp.companion_mission_time_ret = calendar::before_time_starts;
+    for( size_t i = 0; i < comp.companion_mission_inv.size(); i++ ) {
+        for( const auto &it : comp.companion_mission_inv.const_stack( i ) ) {
+            g->m.add_item_or_charges( g->u.posx(), g->u.posy(), it );
+        }
+    }
+    comp.companion_mission_inv.clear();
     // npc *may* be active, or not if outside the reality bubble
     g->reload_npcs();
 }
 
-std::vector<std::shared_ptr<npc>> talk_function::companion_list( const npc &p, const std::string &id )
+std::vector<std::shared_ptr<npc>> talk_function::companion_list( const npc &p, const std::string &id, bool contains )
 {
     std::vector<std::shared_ptr<npc>> available;
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( p.posx(), p.posy() ) );
+    std::string mission_format = string_format( "%s(%d,%d,%d)%s", p.companion_mission_role_id, omt_pos.x, omt_pos.y, p.posz(), id );
```

Comment:
You really, really, really shouldn't use complex formatted strings for anything but player display.
This will get painful fast (if it isn't as painful as it looks yet) and is certain to get painful if you need to change the format at some point (for example, adding new variables).

Instead you should wrap the companion missions in a `struct` and have this `struct` carry all the fields necessary to describe the mission.
For example,
```
struct npc_companion_mission {
    std::string mission_id;
    tripoint position;
    std::string role_id;
};
```